### DISCUSSION TITLE
Cutadapt --discard-untrimmed

### DISCUSSION THREAD

**Discussion by marchoeppner** [bio-raum/FooDMe2#35] [General]
I am wondering how this command line flag works. From the documentation, it I would assume that any read in which a match to any of the primer sequences was found, will be forwarded. For our use case, this is perhaps not ideal - when we request --cutadapt_trim_3p

Just looking at the ONT data, reads that have a primer at the 5' end, but not at the 3' end, will also be considered as "pass". And that is something we probably do not want. We could go back to the original cutadapt code with the pipes - and have both the forward and reverse search use --discard-untrimmed? Although we would lose  the ability to produce the JSON file. 

Thoughts?

**Comment by gregdenay** [comment]
I think that was one of the reason for the somewhat complicated implementation of cutadapt in FooDMe 1:

```
 # Simple case only 5p trimming
        if [[ {params.primer_3p} == False ]]
        then
            cutadapt {input.r1} \
                {input.r2} \
                -o {output.r1} \
                -p {output.r2} \
                -g file:{params.primers} \
                -G file:{params.primers} \
                --untrimmed-output {output.trash_R1_5p} \
                --untrimmed-paired-output {output.trash_R2_5p} \
                --error-rate {params.error_rate} \
                2>&1 > {log}
            touch {output.trash_R1_3p}
            touch {output.trash_R2_3p}

        # in case trimming of 3p is also nescessary
        else
            cutadapt --interleaved \
                {input.r1} \
                {input.r2} \
                -g file:{params.primers} \
                -G file:{params.primers} \
                --untrimmed-output {output.trash_R1_5p} \
                --untrimmed-paired-output {output.trash_R2_5p} \
                --error-rate {params.error_rate} \
            2>> {log} \
            | cutadapt --interleaved \
                -o {output.r1} \
                -p {output.r2} \
                -a file:{input.primers_rc} \
                -A file:{input.primers_rc} \
                --untrimmed-output {output.trash_R1_3p} \
                --untrimmed-paired-output {output.trash_R2_3p} \
                --error-rate {params.error_rate} \
                - \
            2>&1 >> {log}
        fi
```

We could also get a JSON for each step and then add a small script to merge the info, or extract the important info from the log and emulate a cutadapt JSON for multiQC.

**Comment by marchoeppner** [comment]
After reading the documentation again, I thinking that we may need to use "linked adapters" to get the kind of behavior we are after. But I am not sure exactly how that would work for our setting where we have fasta files with any number of sequences rather than a specific fwd<->rev sequence pair

https://cutadapt.readthedocs.io/en/stable/guide.html#linked-adapters

Thoughts?

**Reply by gregdenay** [reply]
I had considered this, it would be in throery quite easy to add a script to also generate all pariwise combination of forward and (reverse-complemt) reverse primers.

But this would mean potentially a large number of combinations for primers with lots of ambiguous nulceotides, which may unnescessarily bloat the cutadapt step.

As an example for the COI primers:
- fwd: GGWACWGGWTGAACWGTWTAYCCYCC -> 128 unambiguous combinations
- rev: TAIACYTCIGGRTGICCRAARAAYCA -> 2048 unambigou combinations (!) 
For a total of 262.144 pairs

That seems absolutely unreasonnable.

I think what we discussed this week, reimplementing the pipe approach and generating a custom table for MQC, although not the most elegant approach is the most flexible and efficient.

**Reply by marchoeppner** [reply]
Yea - I tried to explain the issue to Marcel:
https://github.com/marcelm/cutadapt/issues/806

Apparently Cutadapt understands IUPAC - so the whole disambiguation may not be necessary (although the revese complementing bit may be a problem..). 

Probably faster to just do what we discussed ;)

**Reply by gregdenay** [reply]
mmm i wasn't aware of cutadapt understanding IUPAC, that's good news.
That means we would have to generate all possible pairs of different primers and rc?
So n(n-1) combinations where n is the number of primers in the fasta file?
That seems doable, I wouldn't expect to have more than 4 to 6 primers at the same time.

Worth a test a the very least.

NB: need a bash/sed line to replace I by N

**Reply by marchoeppner** [reply]
Sounds like that may work. I am testing the pipe-based approach and how to sum up the two JSONs. 

As for automatically fixing Inosines - hm, maybe better not to do that. The disambiguation script will now throw a helpful error so people can fix the problem themselves. It's always a bit dangerious to try and "fix" stuff inside the pipeline since it is not entirely transparent to the user.

**Reply by marchoeppner** [reply]
So, I have now a first "prototype" of a somewhat "hacked" cutadapt reporting. Basically, if 3' and 5' trimming is requested, it will do the "pipe" based processing, and each step will emit its own JSON. A script then reads these JSONs and summarizes only the relevant information and generates a MultiQC-compliant result that more-or-less emulates the original Cutadapt bargraph in MultiQC. 

To not break too much stuff, this updated Cutadapt module is currently stored under modules/cutadapt_test ; should be easy enough to revert to the original implementation if we find a better way of doing this.

**Reply by gregdenay** [reply]
Just ran the benchmark, same results as before 👍
